File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.java
import java.util.stream.*;
import java.util.*;
class Solution {
    public int solution(String str1, String str2) {
        Map<String, Integer> map1 = makeMap(str1);
        Map<String, Integer> map2 = makeMap(str2);

        HashSet<String> totalKey = new HashSet<>();

        totalKey.addAll(map1.keySet());
        totalKey.addAll(map2.keySet());

        int gyo = 0;
        int hap = 0;

        for (String key : totalKey) {
            boolean isA = map1.containsKey(key);
            boolean isB = map2.containsKey(key);

            if (isA && isB) {
                Integer a = map1.get(key);
                Integer b = map2.get(key);
                gyo += Math.min(a, b);
                hap += Math.max(a, b);
                continue;
            }

            Integer a = map1.getOrDefault(key, 0);
            Integer b = map2.getOrDefault(key, 0);

            hap += Math.max(a, b);
        }
        int result = hap == 0 ? 65536 : (int) ((double) gyo / hap * 65536);
        return result;
    }
    public Map<String, Integer> makeMap(String str) {
        str = str.toLowerCase();

        char[] chars = str.toCharArray();
        HashMap<String, Integer> map = new HashMap<>();

        char[] makeStr = new char[2];

        for (int i = 0; i < chars.length - 1; i++) {
            //둘다 영어 이면
            if (
                    'a' <= chars[i] && chars[i] <= 'z' &&
                            'a' <= chars[i + 1] && chars[i + 1] <= 'z'

            ) {
                makeStr[0] = chars[i];
                makeStr[1] = chars[i + 1];

                String input = String.valueOf(makeStr);

                map.put(input, map.getOrDefault(input, 0) + 1);
            }
        }

        return map;
    }
}

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.py
import math


def isAlpha(s):
    return ord("a") <= ord(s) <= ord("z") or ord("A") <= ord(s) <= ord("Z")


def counter(str1):
    dic = {}
    for i in range(len(str1) - 1):
        if isAlpha(str1[i]) and isAlpha(str1[i + 1]):
            s = str1[i].lower() + str1[i + 1].lower()
            dic[s] = dic.get(s, 0) + 1

    return dic


def 교(str1, str2):
    c1 = counter(str1)
    c2 = counter(str2)

    c3 = {}
    for k, _ in c1.items():
        if k in c2:
            c3[k] = min(c1[k], c2[k])

    return sum(c3.values())


def 합(str1, str2):
    c1 = counter(str1)
    c2 = counter(str2)

    for k in c2.keys():
        c1[k] = max(c1.get(k, 0), c2[k])

    return sum(c1.values())


def solution(str1, str2):
    a = 교(str1, str2)
    b = 합(str1, str2)
    return math.floor((a / b if b != 0 else 1) * 65536)
